File: indian-options-vibe/frontend/lib/marketSession.ts
export type MarketSessionStatus = {
  isOpen: boolean;
  label: "OPEN" | "CLOSED";
  reason: string;
  istTime: string;
};

function getIstParts(date = new Date()) {
  const parts = new Intl.DateTimeFormat("en-IN", {
    timeZone: "Asia/Kolkata",
    weekday: "short",
    hour: "2-digit",
    minute: "2-digit",
    hour12: false,
  }).formatToParts(date);

  const get = (type: string) => parts.find((part) => part.type === type)?.value || "";

  return {
    weekday: get("weekday"),
    hour: Number(get("hour")),
    minute: Number(get("minute")),
  };
}

export function getMarketSessionStatus(date = new Date()): MarketSessionStatus {
  const { weekday, hour, minute } = getIstParts(date);
  const minutes = hour * 60 + minute;

  const marketOpen = 9 * 60 + 15;
  const marketClose = 15 * 60 + 30;
  const isWeekend = weekday === "Sat" || weekday === "Sun";

  const istTime = new Intl.DateTimeFormat("en-IN", {
    timeZone: "Asia/Kolkata",
    weekday: "short",
    hour: "2-digit",
    minute: "2-digit",
    second: "2-digit",
    hour12: false,
  }).format(date);

  if (isWeekend) {
    return {
      isOpen: false,
      label: "CLOSED",
      reason: "Weekend. NSE live permission is blocked.",
      istTime,
    };
  }

  if (minutes < marketOpen) {
    return {
      isOpen: false,
      label: "CLOSED",
      reason: "Before NSE market open. Live permission is blocked.",
      istTime,
    };
  }

  if (minutes > marketClose) {
    return {
      isOpen: false,
      label: "CLOSED",
      reason: "After NSE market close. Live permission is blocked.",
      istTime,
    };
  }

  return {
    isOpen: true,
    label: "OPEN",
    reason: "NSE market session is open.",
    istTime,
  };
}

File: indian-options-vibe/backend/quant/market_session.py
from datetime import datetime, time
from zoneinfo import ZoneInfo


IST = ZoneInfo("Asia/Kolkata")


def get_market_session_status() -> dict:
    now = datetime.now(IST)
    weekday = now.weekday()  # Monday=0, Sunday=6

    market_open_time = time(9, 15)
    market_close_time = time(15, 30)

    is_weekday = weekday < 5
    is_open_time = market_open_time <= now.time() <= market_close_time
    is_open = is_weekday and is_open_time

    if not is_weekday:
        reason = "Weekend. NSE live trading is closed."
    elif now.time() < market_open_time:
        reason = "Before market open. NSE live trading starts at 09:15 IST."
    elif now.time() > market_close_time:
        reason = "After market close. NSE live trading ended at 15:30 IST."
    else:
        reason = "Market session is open."

    return {
        "timezone": "Asia/Kolkata",
        "now_ist": now.isoformat(),
        "is_weekday": is_weekday,
        "is_open_time": is_open_time,
        "is_open": is_open,
        "reason": reason,
    }
